## Data Collection & Preparation

* ``Pandas`` for managing the data
* ``NumPy`` for math operations
* ``Sklearn`` for Machine Learning
* ``Seaborn`` for advanced visualization
* ``Matplotlib`` for additional plotting

In [1213]:
# Import libraries 

import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler, PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.metrics import mean_squared_error, r2_score
import warnings
warnings.filterwarnings("ignore", category=UserWarning) 
%matplotlib inline

In [1214]:
#read data files

cs = pd.read_csv("charging_sessions.csv")
wba = pd.read_csv("weather_burbank_airport.csv")

In [1215]:
#print the first 5 rows of the charging_sessions dataset
cs.head()

Unnamed: 0                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   

              disconnectTime           doneChargingTime  kWhDelivered  \
0  2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1  2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2  2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3  2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4  2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   

                                sessionID  siteID  spaceID    stationID  \
0  1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1  1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2  1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3  1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4  1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   

              timezone  userID  \
0  America/Los_Angeles   194.0   
1  America/Los_Angeles  4275.0   
2  America/Los_Angeles   344.0   
3  America/Los_Angeles  1117.0   
4  America/Los_Angeles   334.0   

                                          userInputs  
0  [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1  [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3  [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4  [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...

In [1216]:
#print out the first 5 rows of the weather burbank airport dataset
wba.head()

city            timestamp  temperature  cloud_cover  \
0  Burbank  2018-01-01 08:53:00          9.0         33.0   
1  Burbank  2018-01-01 09:53:00          9.0         33.0   
2  Burbank  2018-01-01 10:53:00          9.0         21.0   
3  Burbank  2018-01-01 11:53:00          9.0         29.0   
4  Burbank  2018-01-01 12:53:00          8.0         33.0   

  cloud_cover_description  pressure  windspeed  precipitation  \
0                    Fair    991.75        9.0            0.0   
1                    Fair    992.08        0.0            0.0   
2                    Haze    992.08        0.0            0.0   
3           Partly Cloudy    992.08        0.0            0.0   
4                    Fair    992.08        0.0            0.0   

   felt_temperature  
0               8.0  
1               9.0  
2               9.0  
3               9.0  
4               8.0

In [1217]:
#print out the columns of charging sessions dataset
cs.columns

Index(['Unnamed: 0', 'id', 'connectionTime', 'disconnectTime',
       'doneChargingTime', 'kWhDelivered', 'sessionID', 'siteID', 'spaceID',
       'stationID', 'timezone', 'userID', 'userInputs'],
      dtype='object')

In [1218]:
#print out the columns of weather burbank airport dataset
wba.columns

Index(['city', 'timestamp', 'temperature', 'cloud_cover',
       'cloud_cover_description', 'pressure', 'windspeed', 'precipitation',
       'felt_temperature'],
      dtype='object')

In [1219]:
#print out the tail of cs dataset to see how many entries there are
cs.tail()

Unnamed: 0                        id             connectionTime  \
66445       10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446       10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447       10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448       10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449       10087  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN

In [1220]:
#print out the tail of cs dataset to see how many entries there are
wba.tail()

city            timestamp  temperature  cloud_cover  \
29239  Burbank  2021-01-01 03:53:00         13.0         33.0   
29240  Burbank  2021-01-01 04:53:00         12.0         33.0   
29241  Burbank  2021-01-01 05:53:00         12.0         33.0   
29242  Burbank  2021-01-01 06:53:00         11.0         33.0   
29243  Burbank  2021-01-01 07:53:00         10.0         33.0   

      cloud_cover_description  pressure  windspeed  precipitation  \
29239                    Fair    986.81        0.0            0.0   
29240                    Fair    986.81       11.0            0.0   
29241                    Fair    987.47        9.0            0.0   
29242                    Fair    987.14       13.0            0.0   
29243                    Fair    987.80        6.0            0.0   

       felt_temperature  
29239              13.0  
29240              12.0  
29241              12.0  
29242              11.0  
29243              10.0

### We will focus on the charging_sessions dataset for a while

In [1221]:
#print information about cs dataset
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


We can see that this dataset has an unnamed column that represents the entry number. For better readability, we will rename this column.

In [1222]:
#rename the unnamed column
cs.rename( columns={'Unnamed: 0':'number'}, inplace=True )

In [1223]:
#check if the column was renamed
cs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66450 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            66450 non-null  int64  
 1   id                66450 non-null  object 
 2   connectionTime    66450 non-null  object 
 3   disconnectTime    66450 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      66450 non-null  float64
 6   sessionID         66450 non-null  object 
 7   siteID            66450 non-null  int64  
 8   spaceID           66450 non-null  object 
 9   stationID         66450 non-null  object 
 10  timezone          66450 non-null  object 
 11  userID            49187 non-null  float64
 12  userInputs        49187 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.6+ MB


## Working with missing values

We are going to follow some steps to work with the missing data. First we will Identify the missing data and then deal with it in order to start working with the dataset.


**Detecting missing numerical data**

From looking at the count of every column, we can see that the "doneChargingTime" has missing values. The missing values in the "userID" and "userInputs" columns are because the user was not registered, so we can ignore them for now.

In [1224]:
#We use the next function to identify missing values:
missing_data = cs.isnull()
missing_data

number     id  connectionTime  disconnectTime  doneChargingTime  \
0       False  False           False           False             False   
1       False  False           False           False             False   
2       False  False           False           False             False   
3       False  False           False           False             False   
4       False  False           False           False             False   
...       ...    ...             ...             ...               ...   
66445   False  False           False           False             False   
66446   False  False           False           False             False   
66447   False  False           False           False             False   
66448   False  False           False           False             False   
66449   False  False           False           False             False   

       kWhDelivered  sessionID  siteID  spaceID  stationID  timezone  userID  \
0             False      False   False    False      False     False   False   
1             False      False   False    False      False     False   False   
2             False      False   False    False      False     False   False   
3             False      False   False    False      False     False   False   
4             False      False   False    False      False     False   False   
...             ...        ...     ...      ...        ...       ...     ...   
66445         False      False   False    False      False     False   False   
66446         False      False   False    False      False     False   False   
66447         False      False   False    False      False     False   False   
66448         False      False   False    False      False     False   False   
66449         False      False   False    False      False     False    True   

       userInputs  
0           False  
1           False  
2           False  
3           False  
4           False  
...           ...  
66445       False  
66446       False  
66447       False  
66448       False  
66449        True  

[66450 rows x 13 columns]

The output we receive is a boolean value that indicates whether the value is in fact missing data.

We use next another function which will quicky give us the number of missing values in each column.
True represents a missing value and False represents that there is a value un the dataset.

In [1225]:
for column in missing_data.columns.values.tolist():
    print(column)
    print(missing_data[column].value_counts())
    print("")

number
number
False    66450
Name: count, dtype: int64

id
id
False    66450
Name: count, dtype: int64

connectionTime
connectionTime
False    66450
Name: count, dtype: int64

disconnectTime
disconnectTime
False    66450
Name: count, dtype: int64

doneChargingTime
doneChargingTime
False    62362
True      4088
Name: count, dtype: int64

kWhDelivered
kWhDelivered
False    66450
Name: count, dtype: int64

sessionID
sessionID
False    66450
Name: count, dtype: int64

siteID
siteID
False    66450
Name: count, dtype: int64

spaceID
spaceID
False    66450
Name: count, dtype: int64

stationID
stationID
False    66450
Name: count, dtype: int64

timezone
timezone
False    66450
Name: count, dtype: int64

userID
userID
False    49187
True     17263
Name: count, dtype: int64

userInputs
userInputs
False    49187
True     17263
Name: count, dtype: int64



We can notice that the columns with missing values are: 

- "doneChargingTime": 4088 missing data 
- "userID": 17263 missing data 
- "userInputs": 17263 missing data 

In [1226]:
#drop all rows containing null values
cs.dropna(axis=0, inplace=False)

number                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...       ...                       ...                        ...   
66444   10082  5d574ad2f9af8b4c10c03651  2019-07-31 14:50:17+00:00   
66445   10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446   10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447   10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448   10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66444  2019-08-01 01:01:18+00:00  2019-07-31 20:29:24+00:00        27.174   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66444  1_1_191_784_2019-07-31 14:50:17.037367       1  AG-4F40  1-1-191-784   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66444  America/Los_Angeles   448.0   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66444  [{'WhPerMile': 200, 'kWhRequested': 28.0, 'mil...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  

[46008 rows x 13 columns]

In [1227]:
#check how many rows are getting dropped
len(cs)-len(cs.dropna())

20442

This number is higher than expected, but because this factors in the null values from userID and userInputs(unregistered users or users with no Input), we want so specifically look for missing values in the "doneChargingTime"-column

In [1228]:
len(cs)-len(cs.dropna(subset = ['doneChargingTime']))

4088

This number tells us that there are 4088 sessions with no timestamp for when the EV got done charging, meaning that it didn't fully charge. Let's look at this data.

In [1229]:
#display records with 'doneCharging = Null
cs_notdone = cs[cs['doneChargingTime'].isnull()]
cs_notdone

number                        id             connectionTime  \
3756     3756  5ea3848df9af8b428bf97225  2020-04-08 16:36:53+00:00   
3757     3757  5ea4d60df9af8b46573f1ee7  2020-04-09 13:23:52+00:00   
3758     3758  5ea4d60df9af8b46573f1ee8  2020-04-09 13:40:03+00:00   
3759     3759  5ea4d60df9af8b46573f1ee9  2020-04-09 14:01:53+00:00   
3760     3760  5ea4d60df9af8b46573f1eea  2020-04-09 15:01:06+00:00   
...       ...                       ...                        ...   
60411    4049  5cb3d2a8f9af8b4551261e75  2019-03-30 00:32:10+00:00   
61097    4735  5cc8e891f9af8b45d975b819  2019-04-14 16:19:12+00:00   
61103    4741  5cca3a54f9af8b49aaa4cbac  2019-04-15 13:00:59+00:00   
61117    4755  5cca3a54f9af8b49aaa4cbba  2019-04-15 13:52:45+00:00   
65530    9168  5d3f8f96f9af8b7f97666a55  2019-07-13 20:13:18+00:00   

                  disconnectTime doneChargingTime  kWhDelivered  \
3756   2020-04-09 03:43:29+00:00              NaN        25.183   
3757   2020-04-09 17:59:02+00:00              NaN        14.136   
3758   2020-04-09 21:43:22+00:00              NaN        14.203   
3759   2020-04-09 16:33:40+00:00              NaN        13.041   
3760   2020-04-09 23:38:08+00:00              NaN         5.797   
...                          ...              ...           ...   
60411  2019-03-30 02:23:14+00:00              NaN        11.148   
61097  2019-04-14 20:38:39+00:00              NaN        11.678   
61103  2019-04-15 13:55:25+00:00              NaN         3.343   
61117  2019-04-15 14:47:01+00:00              NaN         4.558   
65530  2019-07-13 20:37:40+00:00              NaN         2.218   

                                    sessionID  siteID  spaceID    stationID  \
3756   1_1_194_822_2020-04-08 16:36:53.316264       1  AG-1F12  1-1-194-822   
3757   1_1_193_827_2020-04-09 13:23:51.689520       1  AG-1F02  1-1-193-827   
3758   1_1_178_823_2020-04-09 13:39:55.115497       1  AG-1F08  1-1-178-823   
3759   1_1_178_828_2020-04-09 14:01:53.467693       1  AG-1F10  1-1-178-828   
3760   1_1_193_820_2020-04-09 15:01:06.343454       1  AG-1F04  1-1-193-820   
...                                       ...     ...      ...          ...   
60411  1_1_193_829_2019-03-30 00:32:10.128651       1  AG-1F03  1-1-193-829   
61097  1_1_193_816_2019-04-14 16:19:11.864262       1  AG-1F05  1-1-193-816   
61103  1_1_193_819_2019-04-15 13:00:58.722284       1  AG-1F06  1-1-193-819   
61117  1_1_179_800_2019-04-15 13:52:44.693153       1  AG-3F32  1-1-179-800   
65530  1_1_193_825_2019-07-13 20:13:18.490314       1  AG-1F01  1-1-193-825   

                  timezone  userID  \
3756   America/Los_Angeles   507.0   
3757   America/Los_Angeles   419.0   
3758   America/Los_Angeles  5065.0   
3759   America/Los_Angeles   651.0   
3760   America/Los_Angeles   466.0   
...                    ...     ...   
60411  America/Los_Angeles   194.0   
61097  America/Los_Angeles   364.0   
61103  America/Los_Angeles   651.0   
61117  America/Los_Angeles   194.0   
65530  America/Los_Angeles    76.0   

                                              userInputs  
3756   [{'WhPerMile': 250, 'kWhRequested': 75.0, 'mil...  
3757   [{'WhPerMile': 400, 'kWhRequested': 40.0, 'mil...  
3758   [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
3759   [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
3760   [{'WhPerMile': 300, 'kWhRequested': 6.0, 'mile...  
...                                                  ...  
60411  [{'WhPerMile': 400, 'kWhRequested': 24.0, 'mil...  
61097  [{'WhPerMile': 400, 'kWhRequested': 56.0, 'mil...  
61103  [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
61117  [{'WhPerMile': 400, 'kWhRequested': 12.0, 'mil...  
65530  [{'WhPerMile': 301, 'kWhRequested': 9.03, 'mil...  

[4088 rows x 13 columns]

### Dealing with missing values

When dealing with missing values, we can either **eliminate** them from the dataset or **impute** the null values with estimates. Because the missing data points are timestamps, it's hard (**to impute**?) them because they simply do not exist. The EV never got done charging, so estimating a value doesn't make any sense. And because our job is to optimize utilization, looking at data points where the EV never fully charged(and thus didn't stay plugged in without charging and taking up charging space) is not adequate. That is why we will drop these data points.

In [1230]:
#we take every row where "doneChargingTime" is not null into our new dataset cs_clean
cs_clean = cs[cs['doneChargingTime'].notna()]

In [1231]:
cs_clean.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62362 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            62362 non-null  int64  
 1   id                62362 non-null  object 
 2   connectionTime    62362 non-null  object 
 3   disconnectTime    62362 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      62362 non-null  float64
 6   sessionID         62362 non-null  object 
 7   siteID            62362 non-null  int64  
 8   spaceID           62362 non-null  object 
 9   stationID         62362 non-null  object 
 10  timezone          62362 non-null  object 
 11  userID            46008 non-null  float64
 12  userInputs        46008 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.7+ MB


In [1232]:
#We can se now that the row "doneChargingTime" has not missing data:

missing_data2 = cs_clean.isnull()
for column in missing_data2.columns.values.tolist():
    print(column)
    print(missing_data2[column].value_counts())
    print("")

number
number
False    62362
Name: count, dtype: int64

id
id
False    62362
Name: count, dtype: int64

connectionTime
connectionTime
False    62362
Name: count, dtype: int64

disconnectTime
disconnectTime
False    62362
Name: count, dtype: int64

doneChargingTime
doneChargingTime
False    62362
Name: count, dtype: int64

kWhDelivered
kWhDelivered
False    62362
Name: count, dtype: int64

sessionID
sessionID
False    62362
Name: count, dtype: int64

siteID
siteID
False    62362
Name: count, dtype: int64

spaceID
spaceID
False    62362
Name: count, dtype: int64

stationID
stationID
False    62362
Name: count, dtype: int64

timezone
timezone
False    62362
Name: count, dtype: int64

userID
userID
False    46008
True     16354
Name: count, dtype: int64

userInputs
userInputs
False    46008
True     16354
Name: count, dtype: int64



The data is now clean and we can work with it!

In [1233]:
#testing grounds for descriptive analysis
cs_test = cs_clean

In [1234]:
cs_test.info()

<class 'pandas.core.frame.DataFrame'>
Index: 62362 entries, 0 to 66449
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   number            62362 non-null  int64  
 1   id                62362 non-null  object 
 2   connectionTime    62362 non-null  object 
 3   disconnectTime    62362 non-null  object 
 4   doneChargingTime  62362 non-null  object 
 5   kWhDelivered      62362 non-null  float64
 6   sessionID         62362 non-null  object 
 7   siteID            62362 non-null  int64  
 8   spaceID           62362 non-null  object 
 9   stationID         62362 non-null  object 
 10  timezone          62362 non-null  object 
 11  userID            46008 non-null  float64
 12  userInputs        46008 non-null  object 
dtypes: float64(2), int64(2), object(9)
memory usage: 6.7+ MB


In [1235]:
# Regarding 2(c):
# To understand which of the two sites is public and which is private, 
# we start by first selecting all rows by their siteID, starting with siteID==1

cs_test.loc[cs_test['siteID']==1,:]

number                        id             connectionTime  \
0           0  5e23b149f9af8b5fe4b973cf  2020-01-02 13:08:54+00:00   
1           1  5e23b149f9af8b5fe4b973d0  2020-01-02 13:36:50+00:00   
2           2  5e23b149f9af8b5fe4b973d1  2020-01-02 13:56:35+00:00   
3           3  5e23b149f9af8b5fe4b973d2  2020-01-02 13:59:58+00:00   
4           4  5e23b149f9af8b5fe4b973d3  2020-01-02 14:00:01+00:00   
...       ...                       ...                        ...   
66445   10083  5d574ad2f9af8b4c10c03652  2019-07-31 18:08:04+00:00   
66446   10084  5d574ad2f9af8b4c10c03653  2019-07-31 18:40:41+00:00   
66447   10085  5d574ad2f9af8b4c10c03654  2019-07-31 19:04:40+00:00   
66448   10086  5d574ad2f9af8b4c10c03655  2019-07-31 19:19:47+00:00   
66449   10087  5d574ad2f9af8b4c10c03656  2019-07-31 19:21:47+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
0      2020-01-02 19:11:15+00:00  2020-01-02 17:31:35+00:00        25.016   
1      2020-01-02 22:38:21+00:00  2020-01-02 20:18:05+00:00        33.097   
2      2020-01-03 00:39:22+00:00  2020-01-02 16:35:06+00:00         6.521   
3      2020-01-02 16:38:39+00:00  2020-01-02 15:18:45+00:00         2.355   
4      2020-01-02 22:08:40+00:00  2020-01-02 18:17:30+00:00        13.375   
...                          ...                        ...           ...   
66445  2019-07-31 23:29:18+00:00  2019-07-31 23:30:18+00:00        28.787   
66446  2019-08-01 00:59:42+00:00  2019-07-31 21:44:23+00:00         7.787   
66447  2019-07-31 22:44:22+00:00  2019-07-31 22:45:21+00:00        11.274   
66448  2019-08-01 00:34:51+00:00  2019-07-31 21:25:30+00:00        11.589   
66449  2019-07-31 22:00:04+00:00  2019-07-31 20:51:34+00:00         0.897   

                                    sessionID  siteID  spaceID    stationID  \
0      1_1_179_810_2020-01-02 13:08:53.870034       1  AG-3F30  1-1-179-810   
1      1_1_193_825_2020-01-02 13:36:49.599853       1  AG-1F01  1-1-193-825   
2      1_1_193_829_2020-01-02 13:56:35.214993       1  AG-1F03  1-1-193-829   
3      1_1_193_820_2020-01-02 13:59:58.309319       1  AG-1F04  1-1-193-820   
4      1_1_193_819_2020-01-02 14:00:00.779967       1  AG-1F06  1-1-193-819   
...                                       ...     ...      ...          ...   
66445  1_1_179_809_2019-07-31 18:08:04.432654       1  AG-3F27  1-1-179-809   
66446  1_1_179_810_2019-07-31 18:40:40.900203       1  AG-3F30  1-1-179-810   
66447  1_1_191_795_2019-07-31 19:04:40.098273       1  AG-4F51  1-1-191-795   
66448  1_1_191_778_2019-07-31 19:19:46.919358       1  AG-4F43  1-1-191-778   
66449  1_1_178_817_2019-07-31 19:21:46.727697       1  AG-1F09  1-1-178-817   

                  timezone  userID  \
0      America/Los_Angeles   194.0   
1      America/Los_Angeles  4275.0   
2      America/Los_Angeles   344.0   
3      America/Los_Angeles  1117.0   
4      America/Los_Angeles   334.0   
...                    ...     ...   
66445  America/Los_Angeles   393.0   
66446  America/Los_Angeles   220.0   
66447  America/Los_Angeles  1974.0   
66448  America/Los_Angeles   942.0   
66449  America/Los_Angeles     NaN   

                                              userInputs  
0      [{'WhPerMile': 250, 'kWhRequested': 25.0, 'mil...  
1      [{'WhPerMile': 280, 'kWhRequested': 70.0, 'mil...  
2      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
3      [{'WhPerMile': 400, 'kWhRequested': 8.0, 'mile...  
4      [{'WhPerMile': 400, 'kWhRequested': 16.0, 'mil...  
...                                                  ...  
66445  [{'WhPerMile': 240, 'kWhRequested': 31.2, 'mil...  
66446  [{'WhPerMile': 333, 'kWhRequested': 6.66, 'mil...  
66447  [{'WhPerMile': 333, 'kWhRequested': 19.98, 'mi...  
66448  [{'WhPerMile': 275, 'kWhRequested': 22.0, 'mil...  
66449                                                NaN  

[33004 rows x 13 columns]

In [1236]:
# Now we select all rows with siteID==2

cs_test.loc[cs_test['siteID']==2,:]

number                        id             connectionTime  \
20179       0  5c412c1df9af8b12cb56c27c  2019-01-01 17:41:45+00:00   
20180       1  5c412c1df9af8b12cb56c27d  2019-01-01 18:09:17+00:00   
20181       2  5c412c1df9af8b12cb56c27e  2019-01-01 18:39:25+00:00   
20182       3  5c412c1df9af8b12cb56c27f  2019-01-01 19:18:53+00:00   
20183       4  5c412c1df9af8b12cb56c280  2019-01-01 21:05:57+00:00   
...       ...                       ...                        ...   
51567    2448  5fff9e9cf9af8b4c2a7970d7  2020-12-29 00:22:25+00:00   
51574    2455  6000f01bf9af8b4e29871065  2020-12-29 22:21:20+00:00   
51581    2462  6002419cf9af8b50514f4c3d  2020-12-30 19:32:38+00:00   
51583    2464  6002419cf9af8b50514f4c3f  2020-12-30 23:02:11+00:00   
51585    2466  6003931bf9af8b5261c8c794  2020-12-31 16:45:14+00:00   

                  disconnectTime           doneChargingTime  kWhDelivered  \
20179  2019-01-01 18:39:21+00:00  2019-01-01 18:40:21+00:00         0.900   
20180  2019-01-02 02:39:32+00:00  2019-01-01 20:16:10+00:00        12.534   
20181  2019-01-01 19:18:49+00:00  2019-01-01 19:19:47+00:00         0.883   
20182  2019-01-01 20:14:07+00:00  2019-01-01 19:59:08+00:00         0.879   
20183  2019-01-02 02:03:02+00:00  2019-01-02 01:59:27+00:00        16.136   
...                          ...                        ...           ...   
51567  2020-12-29 01:32:10+00:00  2020-12-29 01:07:43+00:00         2.092   
51574  2020-12-30 02:37:08+00:00  2020-12-29 23:30:29+00:00         2.324   
51581  2020-12-30 20:34:51+00:00  2020-12-30 20:34:46+00:00         3.068   
51583  2020-12-31 02:53:06+00:00  2020-12-31 00:08:50+00:00         2.244   
51585  2020-12-31 17:30:59+00:00  2020-12-31 17:30:54+00:00         5.289   

                                     sessionID  siteID spaceID     stationID  \
20179  2_39_138_566_2019-01-01 17:41:44.784919       2  CA-512  2-39-138-566   
20180   2_39_79_379_2019-01-01 18:09:16.991864       2  CA-327   2-39-79-379   
20181  2_39_138_566_2019-01-01 18:39:24.566872       2  CA-512  2-39-138-566   
20182  2_39_138_566_2019-01-01 19:18:52.843645       2  CA-512  2-39-138-566   
20183   2_39_79_378_2019-01-01 21:05:56.972890       2  CA-326   2-39-79-378   
...                                        ...     ...     ...           ...   
51567   2_39_125_21_2020-12-29 00:22:24.970035       2  CA-311   2-39-125-21   
51574   2_39_78_365_2020-12-29 22:21:20.198961       2  CA-321   2-39-78-365   
51581   2_39_125_21_2020-12-30 19:32:38.465177       2  CA-311   2-39-125-21   
51583   2_39_78_365_2020-12-30 23:02:11.119521       2  CA-321   2-39-78-365   
51585   2_39_127_19_2020-12-31 16:45:13.554563       2  CA-309   2-39-127-19   

                  timezone  userID  \
20179  America/Los_Angeles     NaN   
20180  America/Los_Angeles   558.0   
20181  America/Los_Angeles     NaN   
20182  America/Los_Angeles     NaN   
20183  America/Los_Angeles  1135.0   
...                    ...     ...   
51567  America/Los_Angeles  1746.0   
51574  America/Los_Angeles  1082.0   
51581  America/Los_Angeles  1746.0   
51583  America/Los_Angeles  1082.0   
51585  America/Los_Angeles  1316.0   

                                              userInputs  
20179                                                NaN  
20180  [{'WhPerMile': 222, 'kWhRequested': 17.76, 'mi...  
20181                                                NaN  
20182                                                NaN  
20183  [{'WhPerMile': 400, 'kWhRequested': 20.0, 'mil...  
...                                                  ...  
51567  [{'WhPerMile': 350, 'kWhRequested': 35.0, 'mil...  
51574  [{'WhPerMile': 290, 'kWhRequested': 5.8, 'mile...  
51581  [{'WhPerMile': 350, 'kWhRequested': 35.0, 'mil...  
51583  [{'WhPerMile': 290, 'kWhRequested': 5.8, 'mile...  
51585  [{'WhPerMile': 283, 'kWhRequested': 28.3, 'mil...  

[29358 rows x 13 columns]

In [1237]:
cs_test["siteID"].value_counts()

siteID
1    33004
2    29358
Name: count, dtype: int64

In [1238]:
# We can see that there are 33004 data entries for the first site and 29358 entries for the second site.
# Next up, we can temporarily split the data set in two parts to ease our upcoming work.
# To do this, we first define two indices to help us drop the rows by the value of siteID.

index_site1 = cs_test[cs_test['siteID']==1].index
index_site2 = cs_test[cs_test['siteID']==2].index

# Now that we have our indices defined, we can drop the rows to split our data set. 
# Note that we use the correct indices so that we drop all rows where siteID==2 for our first set.

cs_test_s1 = cs_test.drop(index_site2)
cs_test_s2 = cs_test.drop(index_site1)


In [1239]:
# We can now verify that the previous step worked as intended. Note the number of data entries for each set.

cs_test_s1.shape

(33004, 13)

In [1240]:
cs_test_s2.shape

(29358, 13)

In [1241]:
# We can determine key differences between the two data sets that will help us understand
# which of the two sites is privately owned and which one is public.

# Let's start off by looking at how many registered users we have in each data set. 
# We do this by checking the sum of null values for both sets.

cs_test_s1.isnull().sum()

number                 0
id                     0
connectionTime         0
disconnectTime         0
doneChargingTime       0
kWhDelivered           0
sessionID              0
siteID                 0
spaceID                0
stationID              0
timezone               0
userID              1481
userInputs          1481
dtype: int64

In [1242]:
# We can now verify that the previous step worked as intended. Note the number of data entries for each set.

In [1243]:
# We can see that there are 1481 unregistered users in our first data set, which contains a total of 33004 entries.
# Therefore, the share of registered users in set 1 is roughly 95,5%. 
(33004-1481)/33004

0.9551266513149921

In [1244]:
# Now, let's do the same for our second set.

cs_test_s2.isnull().sum()

number                  0
id                      0
connectionTime          0
disconnectTime          0
doneChargingTime        0
kWhDelivered            0
sessionID               0
siteID                  0
spaceID                 0
stationID               0
timezone                0
userID              14873
userInputs          14873
dtype: int64

In [1245]:
# In our second set (with a total of 29358 data entries), there are 14873 entries with missing userIDs. 
# Therefore, the share of registered users in this set is roughly 49,3%.

# We can obvserve a significant difference in share of registered users for both data sets, 
# and thus for the two charging sites (Site 1: 95,5%, Site 2: 49,3%).

# Using our domain knowledge, it is reasonable to assume that the majority of consumers with access 
# to a privately owned charging site visit it frequently. This is reinforced by the fact that the privately owned
# site is open to employees of a company. This is in contrast to a public charging site, 
# for which we would expect a significant amount of one-time consumers, making use of public charging sites
# on the go. Additionally, we have reason to assume that customers with access to a private charging site would
# likely have a registered account, as they frequent the site and thus could make use of the benefits that come
# with registering an account. Looking at the stark contrast of registered user share between the two sites, 
# we can confidently assume that Site 1 is the privately owned one, while Site 2 is public.

# Note for our assignment:
# Additional metrics would definitely be of benefit, 
# but I think this one alone is pretty much all we need as the minimum 

In [1246]:
#tests for 2(a) - temporal and seasonal differences

# Since the time values for all datatime objects are in UTC (GMT), we first have to correct the values. 
# The timezone for Los Angeles is GMT-8, so we have to adjust all time values for every data entry by 8 hours.

from datetime import datetime, timedelta

# As an example on how this can be done, let's take the current datetime (which is GMT+1 for us in Germany) and
# subtract 9 hours using timedelta. This will give us the current datetime for LA.

current_time_LA = datetime.now() - timedelta(hours=9)
current_time_LA

datetime.datetime(2024, 1, 6, 16, 21, 55, 423170)

In [1247]:
cs_test2 = cs_test
cs_test2["connectionTime"].info()

<class 'pandas.core.series.Series'>
Index: 62362 entries, 0 to 66449
Series name: connectionTime
Non-Null Count  Dtype 
--------------  ----- 
62362 non-null  object
dtypes: object(1)
memory usage: 3.0+ MB


In [1248]:
# Here we apply the datetime.strptime() function for a single data entry:

cs_test2.loc[0, 'connectionTime'] = datetime.strptime(cs_test2.loc[0, 'connectionTime'], "%Y-%m-%d %H:%M:%S+00:00")

cs_test2.loc[0, 'connectionTime']

datetime.datetime(2020, 1, 2, 13, 8, 54)

In [1249]:
# For testing purposes, let's try to apply this adjustment for a single data entry first, the row number 0.
# Upon running the code, we can notice that the time data is still stored as a String object. 
# We can use the datetime.strptime() method to convert String objects to datetime objects.

cs_test2.loc[0, 'connectionTime'] = cs_test2.loc[0, 'connectionTime'] - timedelta(hours=8)

In [1250]:
cs_test2["connectionTime"].head(5)

0          2020-01-02 05:08:54
1    2020-01-02 13:36:50+00:00
2    2020-01-02 13:56:35+00:00
3    2020-01-02 13:59:58+00:00
4    2020-01-02 14:00:01+00:00
Name: connectionTime, dtype: object

In [1251]:
# We have successfully transformed this single data entry from String object to datatime object.
# Now we can do the same for all data entries in our cs_test2 dataframe

In [1252]:
#In order to do not commit any error, lets first reset the chage we made to the first row:

cs_test2.loc[0, 'connectionTime'] = cs_test2.loc[0, 'connectionTime'] + timedelta(hours=8)
cs_test2.loc[0, 'connectionTime'] = cs_test2.loc[0, 'connectionTime'].strftime('%Y-%m-%d %H:%M:%S+00:00')


In [1253]:
#We now transform all the rows:

for i in range(len(cs_test2)):
    try:
        connectionTime_str = cs_test2.loc[i, 'connectionTime']
        connectionTime_dt = datetime.strptime(connectionTime_str, '%Y-%m-%d %H:%M:%S+00:00')
        cs_test2.loc[i, 'connectionTime'] = connectionTime_dt
    except KeyError:
        # Handle the case where the index is not found
        pass



In [1257]:
#Now we need to change the all datatime objects (which are in UTC) to the timezone for Los Angeles (GMT-8).

for i in range(len(cs_test2)):
    try:
            cs_test2.loc[i, 'connectionTime'] = cs_test2.loc[i, 'connectionTime'] - timedelta(hours=8)
    except KeyError:
        # Handle the case where the index is not found
        pass

In [1169]:
#Now we can see the result:

cs_test2["connectionTime"].head(5)

0    2020-01-02 05:08:54
1    2020-01-02 05:36:50
2    2020-01-02 05:56:35
3    2020-01-02 05:59:58
4    2020-01-02 06:00:01
Name: connectionTime, dtype: object